### ISSUE TITLE
--ntilts does not do anything in backprojection

### ISSUE THREAD

**Issue by michal-g** [ml-struct-bio/cryodrgn#379]
**Describe the bug**
The argument `--ntilts` to the `cryodrgn backproject_voxel` command does not do anything, even when `--tilt` is given, with all tilts being used regardless in the backprojection.

**To Reproduce**
For example, with a command like
```
cryodrgn backproject_voxel particles_from_M.star --tilt --ntilts=20 --lazy --datadir subtiltstacks/128 --ctf ctf.pkl --poses pose.pkl --dose-per-tilt 2.93 --angle-per-tilt 3 -o backproj.128_20.mrc
```

in the logs at the beginning you get:

```
(INFO) (dataset.py) (19-Jun-24 11:10:31) Loaded 757106 tilts for 18466 particles
(INFO) (dataset.py) (19-Jun-24 11:10:31) {41} tilts per particle
```

and then at the end:

```
(INFO) (backproject_voxel.py) (19-Jun-24 12:53:58) fimage 757100
(INFO) (backproject_voxel.py) (19-Jun-24 12:53:58) Backprojected 757106 images in 6231.20s (0.008230s per image)
(INFO) (fsc.py) (19-Jun-24 12:54:00) res @ FSC=0.5: 8.471 ang
(INFO) (fsc.py) (19-Jun-24 12:54:00) res @ FSC=0.143: 7.933 ang
```

Likewise with a command like the following which uses the same inputs but with 10 tilts instead of 20:
```
cryodrgn backproject_voxel particles_from_M.star --tilt --ntilts=10 --lazy --datadir subtiltstacks/128 --ctf ctf.pkl --poses pose.pkl --dose-per-tilt 2.93 --angle-per-tilt 3 -o backproj.128_10.mrc
```

you still get

```
(INFO) (dataset.py) (19-Jun-24 11:12:08) Loaded 757106 tilts for 18466 particles
(INFO) (dataset.py) (19-Jun-24 11:12:08) {41} tilts per particle
```

and then likewise at the end:

```
(INFO) (backproject_voxel.py) (19-Jun-24 12:54:35) fimage 757100
(INFO) (backproject_voxel.py) (19-Jun-24 12:54:35) Backprojected 757106 images in 6169.96s (0.008149s per image)
(INFO) (fsc.py) (19-Jun-24 12:54:36) res @ FSC=0.5: 8.471 ang
(INFO) (fsc.py) (19-Jun-24 12:54:36) res @ FSC=0.143: 7.933 ang
```

**Expected behavior**
In the above two cases, you would expect 184660 and 369320 images being used respectively in the backprojection as `--ntilts` should restrict the backprojection to using only that many of the tilt images available per particle.

**Comment by michal-g** [comment]
Iterating over a dataset's images in `backproject_voxel` is done by creating an iterator of image indices `{0, 1, ..., nimgs}`, and then using these indices to fetch images individually using `ff = data.get_tilt(ii) if args.tilt else data[ii]`:
```
iterator = range(min(args.first, Nimg)) if args.first else range(Nimg)
```

`args.ntilts` is indeed passed to the dataset (`data` above, a `cryodrgn.dataset.TiltSeriesData` object in this case), but `Nimg = data.N` is always the total number of dataset images, so `ntilts` is effectively ignored when creating the iterator!

In 9de3317 we fix this by modifying the iterator to only include image indices from tilts corresponding to the first `ntilts` tilts for each particle:

```
iterator = range(min(args.first, Nimg)) if args.first else range(Nimg)
if args.tilt:
    use_tilts = set(range(args.ntilts))
    iterator = [
        ii for ii in iterator if int(data.tilt_numbers[ii].item()) in use_tilts
    ]
```

Compare this to the way dataset images are iterated over in e.g. `cryodrgn train_vae`, where batches of input data are produced directly by a dataloader object iteration method, which takes the `ntilts` of the underlying dataset into account. We should perhaps create a special iteration method for datasets for this case of producing one image at a time!

The above fix will be made available in the upcoming `v3.3.3` release and in the meantime is available through our beta release channel as of `v3.3.3-b0`:
```
pip install -i https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple/ cryodrgn --pre
```

**Comment by michal-g** [comment]
To confirm that `backproject_voxel` now behaves as expected, we reproduced the result from Figure S2b of our [cryoDRGN-ET preprint](https://www.biorxiv.org/content/10.1101/2023.08.18.553799v1.full.pdf), in which backprojection was used to evaluate the effect of including more tilts on the quality of a homogeneous reconstruction:
![fsc-plot](https://github.com/ml-struct-bio/cryodrgn/assets/1620067/b22d962c-9fad-4fce-8aa9-519416de5fdc)



Thus we see that as in the paper, there is a clear pattern of improved performance with more tilts that hits diminishing returns past ntilts=8!
